### ISSUE TITLE
Init script robustness: Delete startup lock before starting up

### ISSUE THREAD

**Issue by benjaoming** [learningequality/installers#310]
See https://github.com/learningequality/ka-lite/issues/4738

**Comment by benjaoming** [comment]
Except well, maybe it's a bad idea. Example:

```
service ka-lite start
service ka-lite restart
service ka-lite restart
```

If the init script removes the startup lock it's gonna get messy.

**Comment by jamalex** [comment]
Can you clarify? The init cleanup script would be (or could be set to be) run before the ka-lite service ever has a chance to run, no?

**Comment by jamalex** [comment]
And since it only happens at boot, we know that KA Lite must not already be running, so the lock file must be stale?

**Comment by jamalex** [comment]
Ah, but this issue was proposing something different.

**Comment by benjaoming** [comment]
I don't think having a cleanup init script is the way to go... you'd have to also make it run before the other service. This behavior is ill-defined and prompts us to use upstart or system.d AFAIK.

But so I think it's complicated because of this:

We have a startup lock, because essentially we want to exclude another `kalite start` from running while we doing `initialize_kalite`. In case, there's a lock file, we check if the PID stated in the lock file exists. If not, we ignore it and start anyways.

So the problem can only occur when:

1. `kalite start` was killed during initialization and
1. the PID of the lock file is in use.

I think it's fairly unlikely, but I could see it happen in the case where the .kalite folder was always removed and re-initialized since the PID could easily be in use in that case.

As for the example, I guess I don't know if we could have `kalite stop` exiting by means of killing an unresponsive `kalite start` (which would be unresponsive while it's doing its initialization). `kalite stop` will kill an unresponsive process:

https://github.com/learningequality/ka-lite/blob/master/kalitectl.py#L605

**Comment by jamalex** [comment]
> you'd have to also make it run before the other service. This behavior is ill-defined and prompts us to use upstart or system.d

Would there be issues with doing so? We know KA Lite doesn't get started until networking is enabled, but since all we're doing here is deleting a file, couldn't it be at an earlier runlevel?

On the Pi in particular, and especially in the Rachel Pi use case (plug-and-play wifi hotspot), it's much more important that KA Lite always start reliably (since the user likely doesn't have shell access/skill) than to avoid the "starting up twice" situation, which is very unlikely for the same reasons (no shell access). So, suggesting how best to do this for them specifically, even if we don't make it part of upstream KA Lite, seems like a good compromise?

**Comment by benjaoming** [comment]
> Would there be issues with doing so? We know KA Lite doesn't get started until networking is enabled, but since all we're doing here is deleting a file, couldn't it be at an earlier runlevel?

That's runlevel 3. But inside that runlevel, I don't know.

Yes, there's another issue about supporting system.d: #271 

But I don't think we should have 2 init scripts as things are now... especially not if we factor out into several different init standards.

**Comment by jamalex** [comment]
Ok, sounds good -- looks like you have a temp solution for them, and then let's leave a general solution until we support system.d
